mountpoint-s3-client/src/imds_crt_client.rs

```diff
@@ -0,0 +1,109 @@
+    pub fn new() -> Result<Self, error::Error> {
+        let allocator = Allocator::default();
+
+        let mut event_loop_group = EventLoopGroup::new_default(&allocator, None, || {}).unwrap();
+
+        let resolver_options = HostResolverDefaultOptions {
+            max_entries: 8,
+            event_loop_group: &mut event_loop_group,
+        };
+
+        let mut host_resolver = HostResolver::new_default(&allocator, &resolver_options).unwrap();
+
+        let bootstrap_options = ClientBootstrapOptions {
+            event_loop_group: &mut event_loop_group,
+            host_resolver: &mut host_resolver,
+        };
+
+        let client_bootstrap = ClientBootstrap::new(&allocator, &bootstrap_options).unwrap();
+
+        let mut retry_strategy_options = StandardRetryOptions::default(&mut event_loop_group);
+        let retry_strategy = RetryStrategy::standard(&allocator, &retry_strategy_options).unwrap();
+
+        let mut client_config = ImdsClientConfig::new();
+
+        client_config
+            .client_bootstrap(client_bootstrap)
+            .retry_strategy(retry_strategy);
+
+        let imds_client = ImdsClient::new(&allocator, client_config).unwrap();
+
+        Ok(Self {
+            imds_client,
+            event_loop_group,
+            allocator,
+        })
+    }
+
+    /// Query the instnace type mountpoint-s3 running on.
+    pub fn make_instance_type_query(
+        &self,
+        callback: impl FnOnce(Result<String, error::Error>) -> Result<Result<String, error::Error>, ImdsQueryRequestError>
+            + 'static,
+    ) -> Result<ImdsQueryRequest, ImdsQueryRequestError> {
+        let (tx, rx) = oneshot::channel::<Result<Result<String, error::Error>, ImdsQueryRequestError>>();
+        let _ = self.imds_client.get_instance_type(move |result| {
+            let res = callback(result);
+            let _ = tx.send(res);
+        });
+
+        Ok(ImdsQueryRequest { receiver: rx })
+    }
+}
+
+#[derive(Debug)]
+#[pin_project]
+/// ImdsQueryRequest is a data encapsulation wrapper for asnychronous instance metadata query.
+pub struct ImdsQueryRequest {
+    #[pin]
+    receiver: oneshot::Receiver<Result<Result<String, error::Error>, ImdsQueryRequestError>>,
+}
+
+impl Future for ImdsQueryRequest {
+    type Output = Result<Result<String, error::Error>, ImdsQueryRequestError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        this.receiver
+            .poll(cx)
+            .map(|result| result.unwrap_or_else(|err| Err(ImdsQueryRequestError::InternalError(Box::new(err)))))
+    }
+}
+
+/// ImdsQueryRequestError is responsible for hide and unify underlying crt and async error.
+#[derive(Error, Debug)]
+pub enum ImdsQueryRequestError {
+    /// An internal error from within the Imds client. The request may have been sent.
+    #[error("Internal Imds client error")]
+    InternalError(#[source] Box<dyn std::error::Error>),
```

Comment:
This will be really annoying to use without Send and Sync:
```suggestion
    InternalError(#[source] Box<dyn std::error::Error + Send + Sync>),
```

Reply:
fixed